okhttp/src/test/java/okhttp3/CertificatePinnerTest.java

```diff
@@ -189,4 +189,53 @@
     } catch (SSLPeerUnverifiedException expected) {
     }
   }
+
+  @Test public void checkForHostnameWithDoubleAsterisk() throws Exception {
+    CertificatePinner certificatePinner = new CertificatePinner.Builder()
+        .add("**.example.co.uk", certA1Sha256Pin)
+        .build();
+
+    // Should be pinned:
+    try {
+      certificatePinner.check("foo.example.co.uk", certB1.certificate());
+      fail();
+    } catch (SSLPeerUnverifiedException expected) {}
+    try {
+      certificatePinner.check("foo.bar.example.co.uk", certB1.certificate());
+      fail();
+    } catch (SSLPeerUnverifiedException expected) {}
+    try {
+      certificatePinner.check("foo.bar.baz.example.co.uk", certB1.certificate());
+      fail();
+    } catch (SSLPeerUnverifiedException expected) {}
+
+    // Should not be pinned:
+    certificatePinner.check("uk", certB1.certificate());
+    certificatePinner.check("co.uk", certB1.certificate());
+    certificatePinner.check("example.co.uk", certB1.certificate());
+    certificatePinner.check("anotherexample.co.uk", certB1.certificate());
+  }
+
+  @Test public void checkForHostnameWithMultipleWildcardSegments() throws Exception {
+    CertificatePinner certificatePinner = new CertificatePinner.Builder()
+            .add("*.*.example.co.uk", certA1Sha256Pin)
```

Comment:
Worth documenting this as it looks like a very generic feature.  Should *.example.* also work? e.g. companies that register with multiple countries?

Reply:
I don't see any good reason not to support it. I have added a test for wildcard segments that are not at the bottom level.

Reply:
I do not want `.*` at the end. There’s lots of weird new failure modes this introduces.

Reply:
I have made attempts to use a wildcard TLD result in an IllegalArgumentException